src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/TimeSpanConverter.cs

```diff
@@ -0,0 +1,289 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+using System.Diagnostics;
+
+namespace System.Text.Json.Serialization.Converters
+{
+    internal sealed class TimeSpanConverter : JsonConverter<TimeSpan>
+    {
+        private static JsonEncodedText Zero = JsonEncodedText.Encode("PT0S");
+        private const ulong TicksPerYear = TimeSpan.TicksPerDay * 365;
+        private const ulong TicksPerMonth = TimeSpan.TicksPerDay * 30;
```

Comment:
I presume these values are convention specified by the duration spec? I couldn't find anything definitive on the matter.

Reply:
https://github.com/dotnet/corefx/blob/6d723b8e5ae3129c0a94252292322fc19673478f/src/System.Private.Xml/src/System/Xml/Schema/XsdDuration.cs#L229-L236

Reply:
I don't think this makes sense. A duration of 48 hours is not the same as a duration of 2 days per ISO 8601. Example: adding 48 hours to 2021-11-06 07:00 in North Carolina is not the same as adding 2 days to the same date. Likewise, I would not expect adding 30 days to 2020-02-01 being the same as adding 1 month to the same date.

Reply:
> https://github.com/dotnet/corefx/blob/6d723b8e5ae3129c0a94252292322fc19673478f/src/System.Private.Xml/src/System/Xml/Schema/XsdDuration.cs#L229-L236

Just because we're doing it already doesn't mean it is devoid of flaws 😄. Is this backed by the specification?

Reply:
Had no original documents, so can't answer your question, but to correctly handle such things the month and year parts should be avoided, but they still can be received and it will be a valid input. The problem is that it's not compatible with .NET date and time representations.

For example, PostgreSQL has `interval` which stores a duration in three parts: months, days, microseconds. And to handle arithmetic operations properly it deconstructs a `timestamp` into parts, adds an interval to it and then assembles a new `timestamp` again. `timestamp` is a 64 bit integer measuring microseconds.

Unfortunately, `TimeSpan` stores ticks and can't be correctly constructed from any input. For that reason it lacks of `Years` and `Months` properties.

Therefore, while I agree with @watfordgnf that it's a wrong logic and serialization should not include years and months, deserialization of `TimeSpan` will cause issues if these parts are present. Surely, I can just throw a `NotSupportedException` for such cases and it will solve the problem.

Reply:
The `XmlConvert` implementation seems to [avoid `M` and `Y` digits completely](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.Xml/src/System/Xml/Schema/XsdDuration.cs#L121-L122) because of that reason.

That being said, I don't see why we shouldn't support `M` and `Y` digits on deserialization, provided that we follow the ISO8601 specification.

Reply:
As I said, I have no access to paid documents, but I'm mostly sure that it contains the correct logic with arithmetic performed for each component separately. The most popular NPM package handles the logic correctly to by having a special type `Duration` which stores components separately.

Reply:
> The most popular NPM package handles the logic correctly to by having a special type Duration which stores components separately.

Is that duration.js? I presume then that the meaning of months can be interpreted relative to dates it is being added to.

Reply:
It is.

Reply:
I'm not sure ISO 8601 Duration is the correct format for serializing .NET's TimeSpan, as they fundamentally cover different concepts.

Consider how NodaTime breaks these concepts apart:
- https://nodatime.org/2.3.x/api/NodaTime.Duration.html
- https://nodatime.org/2.3.x/api/NodaTime.Period.html

Reply:
It's known that NodaTime is more preferable way to represent time in .NET, but the problem here is a bit different. The serializer somehow should allow transferring data between .NET applications and applications made using different frameworks, and the world isn't .NET centric, so no dictation should be done, but instead a standard should be used.

While it would be nice to see changes in BCL, there's a rule - everything made here is written in stone.

Therefore, let's concentrate on how to change the proposed implementation or seeking for a better format 🙂

Reply:
NodaTime provides a pattern for JSON, but have no idea is it a standard or at least wide spread practice:

https://github.com/nodatime/nodatime/blob/af257385d785f597fc8be67c84f2cf714fbe4203/src/NodaTime/Text/DurationPattern.cs

Reply:
I believe it would be the least confusing for users if TimeSpan was serialized using its native .NET format, rather than a standard it doesn't meet (it represents a subset of it at best). Perhaps not the most convenient when interoperating with other services, but if they expect ISO 8601 durations you couldn't use them reliably anyway (bidirectionally).

Reply:
After conversation with @layomia we have concluded that ISO 8601 durations is not the appropriate format for encoding `TimeSpan`. As @watfordgnf points out `TimeSpan` represents a subset of the standard so deserialization cannot be implemented without substantial sacrifices in precision.

The converter should instead use the standard .NET format for timespan. This has the benefit of being compatible with json.net (which can be an issue with, say, with document databases like cosmos currently using json.net).

Reply:
How should tests look like then? Do the same thing as for `DateTime`?

Reply:
Yeah I think DateTime tests might be a good starting point. Given that the parsing will be pushed out of the json layer the coverage doesn't need to be as extensive.